lib/httplib/httpheaders.go

```diff
@@ -118,43 +134,96 @@ func SetDefaultSecurityHeaders(h http.Header) {
 	h.Set("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
 }
 
-// SetIndexContentSecurityPolicy sets the Content-Security-Policy header for main index.html page
-func SetIndexContentSecurityPolicy(h http.Header, cfg proto.Features) {
-	cspMaps := []map[string]string{
-		defaultContentSecurityPolicy,
-		defaultFontSrc,
-		{"connect-src": "'self' wss:"},
-	}
+func getIndexContentSecurityPolicy(withStripe, withWasm bool) cspMap {
+	cspMaps := []cspMap{defaultContentSecurityPolicy, defaultFontSrc, defaultConnectSrc}
 
-	if cfg.GetCloud() && cfg.GetIsUsageBased() {
+	if withStripe {
 		cspMaps = append(cspMaps, stripeSecurityPolicy)
 	}
 
-	cspString := getContentSecurityPolicyString(cspMaps...)
+	if withWasm {
+		cspMaps = append(cspMaps, wasmSecurityPolicy)
+	}
+
+	return combineCSPMaps(cspMaps...)
+}
+
+// desktopSessionRe is a regex that matches /web/cluster/:clusterId/desktops/:desktopName/:username
+// which is a route to a desktop session that uses WASM.
+var desktopSessionRe = regexp.MustCompile(`^/web/cluster/[^/]+/desktops/[^/]+/[^/]+$`)
+
+var indexCspStringCache = make(map[string]string)
+
+func getIndexContentSecurityPolicyString(cfg proto.Features, urlPath string) string {
+	// Check for result with this cfg and urlPath in cache
+	withStripe := cfg.GetCloud() && cfg.GetIsUsageBased()
+	key := fmt.Sprintf("%v-%v", withStripe, urlPath)
+	if cspString, ok := indexCspStringCache[key]; ok {
+		return cspString
+	}
+
+	// Nothing found in cache, calculate regex and result
+	withWasm := desktopSessionRe.MatchString(urlPath)
+	cspString := getContentSecurityPolicyString(
+		getIndexContentSecurityPolicy(withStripe, withWasm),
+	)
+	// Add result to cache
+	indexCspStringCache[key] = cspString
```

Comment:
This is a data race, no?

Reply:
Yes good catch. The mutex adds a bit of overhead but still an OOM improvement (d2154eb2a79b8f30d2506620dc69448d4ef85be1)

```
$ /opt/homebrew/bin/go test -test.v -timeout 30s -tags desktop_access_rdp -run ^TestSetIndexContentSecurityPolicy$ github.com/gravitational/teleport/lib/httplib 2>&1 | tee out.log
=== RUN   TestSetIndexContentSecurityPolicy
=== PAUSE TestSetIndexContentSecurityPolicy
=== CONT  TestSetIndexContentSecurityPolicy
=== RUN   TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm)
    httplib_test.go:382: Min time: 208ns
    httplib_test.go:383: Max time: 31.834µs
    httplib_test.go:384: Mean time: 324ns
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm)
    httplib_test.go:382: Min time: 209ns
    httplib_test.go:383: Max time: 19.291µs
    httplib_test.go:384: Mean time: 324ns
=== RUN   TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm)
    httplib_test.go:382: Min time: 208ns
    httplib_test.go:383: Max time: 13µs
    httplib_test.go:384: Mean time: 320ns
=== RUN   TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm)
    httplib_test.go:382: Min time: 250ns
    httplib_test.go:383: Max time: 32.708µs
    httplib_test.go:384: Mean time: 395ns
--- PASS: TestSetIndexContentSecurityPolicy (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/default_(no_stripe_or_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_(with_stripe,_no_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_desktop_session_(no_stripe,_with_wasm) (0.00s)
    --- PASS: TestSetIndexContentSecurityPolicy/for_cloud_based_usage_&_desktop_session_(with_stripe,_with_wasm) (0.00s)
PASS
ok  	github.com/gravitational/teleport/lib/httplib	0.429s
```